rpcs3/Emu/Io/usio.cpp

```diff
@@ -565,38 +533,31 @@ void usb_device_usio::interrupt_transfer(u32 buf_size, u8* buf, u32 endpoint, Us
 		if ((buf[0] & USIO_COMMAND_READ) == USIO_COMMAND_READ)
 		{
 			usio_log.trace("UsioRead(Channel: 0x%02X, Register: 0x%04X, Length: 0x%04X)", usio_channel, usio_register, usio_length);
+			response_seek = 0;
+			response.clear();
 			usio_read(usio_channel, usio_register, usio_length);
 			return;
 		}
 
 		// Init and reset commands
-		if ((buf[0] & 0xA0) == 0xA0)
+		if ((buf[0] & USIO_COMMAND_INIT) == USIO_COMMAND_INIT)
 		{
-			const std::array<u8, 2> init_command = {0xA0, 0xF0}; // This kind of command starts with 0xA0, 0xF0 commonly. For example, {0xA0, 0xF0, 0x28, 0x00, 0x00, 0x80}
-			ensure(memcmp(buf, init_command.data(), 2) == 0);
+			//const std::array<u8, 2> init_command = {0xA0, 0xF0}; // This kind of command starts with 0xA0, 0xF0 commonly. For example, {0xA0, 0xF0, 0x28, 0x00, 0x00, 0x80}
+			//ensure(memcmp(buf, init_command.data(), 2) == 0);
 			return;
 		}
 
-		fmt::throw_exception("Received an unexpected command: 0x%02X", buf[0]);
+		usio_log.error("Received an unexpected command: 0x%02X", buf[0]);
 	}
 	case 0x82:
 	{
 		// Read endpoint
-		if (!q_replies.empty())
+		if (response_seek < response.size())
 		{
-			// Sometimes software will outright ignore what usio sends and read with a buffer of 0
-			if (buf_size == 0)
-			{
-				transfer->expected_count = q_replies.front().size();
-				q_replies.pop();
-				break;
-			}
-
-			// Otherwise we expect the buffer to be appropriately sized
-			ensure(buf_size >= q_replies.front().size());
-			memcpy(buf, q_replies.front().data(), q_replies.front().size());
-			transfer->expected_count = q_replies.front().size();
-			q_replies.pop();
+			u32 size = std::min(buf_size, static_cast<u32>(response.size()));
+			memcpy(buf, response.data() + response_seek, size);
```

Comment:
if size == response.size() and response_seek > 0 this will be out of bounds

Reply:
But the code in the block won't be executed when response_seek < response.size().

Reply:
how does this affect what i said xD

Reply:
Just wondering how it can be out of bounds.
response_seek will be set to 0 in next UsioRead action.

Reply:
This is irrelevant. the way you wrote it doesn't make sense.
Change
`u32 size = std::min(buf_size, static_cast<u32>(response.size()));`
to
`const u32 size = std::min(buf_size, static_cast<u32>(response.size() - response_seek));`

Reply:
Got it. Thanks a lot for pointing that out!